File: authentication/src/main/java/authentication/config/security/SecurityConfig.java
package authentication.config.security;

import jakarta.servlet.http.HttpServletResponse;
import lombok.extern.slf4j.Slf4j;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.web.cors.CorsConfiguration;
import org.springframework.web.cors.CorsConfigurationSource;
import org.springframework.web.cors.UrlBasedCorsConfigurationSource;

import java.util.List;

@Configuration
@EnableMethodSecurity
@Slf4j
public class SecurityConfig {

    @Bean
    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {

        return http
                .cors(cors -> cors.configurationSource(corsConfigurationSource()))
                .csrf(csrf -> csrf.disable())
                .authorizeHttpRequests(auth -> auth
                        .requestMatchers(
                                "/actuator/health",
                                "/auth/status")
                        .permitAll()
                        .anyRequest().authenticated()
                )
                .exceptionHandling(ex -> ex
                        .authenticationEntryPoint((req, res, authEx) -> {
                            String reason = authEx.getMessage() == null ? "Unauthorized" : authEx.getMessage();
                            String escapedReason = reason.replace("\"", "\\\"");
                            log.warn("Unauthorized request: method={}, uri={}, reason={}",
                                    req.getMethod(),
                                    req.getRequestURI(),
                                    reason);
                            res.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
                            res.setContentType("application/json;charset=UTF-8");
                            res.getWriter().write("{\"message\": \"Unauthorized\", \"reason\": \"" + escapedReason + "\"}");
                        })
                )
                .build();
    }

    @Bean
    public CorsConfigurationSource corsConfigurationSource() {
        CorsConfiguration configuration = new CorsConfiguration();
        configuration.setAllowedOrigins(List.of(
                "http://localhost:3000",
                "http://localhost:5173",
                "http://localhost:8888",
                "http://localhost:8080",
                "https://dev-pick.com"
        ));
        configuration.setAllowedMethods(List.of("GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS"));
        configuration.setAllowedHeaders(List.of("*"));
        configuration.setAllowCredentials(true);

        UrlBasedCorsConfigurationSource source = new UrlBasedCorsConfigurationSource();
        source.registerCorsConfiguration("/**", configuration);
        return source;
    }
}

File: api/src/main/kotlin/api/global/config/SecurityConfig.kt
package api.global.config

import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.web.SecurityFilterChain
import org.springframework.web.cors.CorsConfiguration
import org.springframework.web.cors.CorsConfigurationSource
import org.springframework.web.cors.UrlBasedCorsConfigurationSource

@Configuration
@EnableWebSecurity
@EnableMethodSecurity
class SecurityConfig {

    @Bean
    fun securityFilterChain(http: HttpSecurity): SecurityFilterChain {
        http
            .cors { }
            .csrf { it.disable() }
            .authorizeHttpRequests { authorize ->
                authorize
                    .requestMatchers("/swagger-ui/**").permitAll()
                    .requestMatchers("/v3/api-docs/**").permitAll()
                    .requestMatchers("/posts", "/posts/**").permitAll()
                    .requestMatchers("/api/posts", "/api/posts/**").permitAll()
                    .requestMatchers("/admin/**").permitAll()
                    .requestMatchers("/categories/**").permitAll()
                    .requestMatchers("/api/categories/**").permitAll()
                    .anyRequest().authenticated()
            }
//            .oauth2ResourceServer { oauth2 ->
//                oauth2.jwt { jwt ->
//                    jwt.jwtAuthenticationConverter(jwtAuthenticationConverter())
//                }
//            }
        return http.build()
    }

//    private fun jwtAuthenticationConverter(): JwtAuthenticationConverter {
//        val grantedAuthoritiesConverter = JwtGrantedAuthoritiesConverter().apply {
//            setAuthoritiesClaimName("cognito:groups")
//            setAuthorityPrefix("ROLE_")
//        }
//        return JwtAuthenticationConverter().apply {
//            setJwtGrantedAuthoritiesConverter(grantedAuthoritiesConverter)
//        }
    @Bean
    fun corsConfigurationSource(): CorsConfigurationSource {
        val configuration = CorsConfiguration().apply {
            allowedOrigins = listOf(
                "http://localhost:3000",
                "http://localhost:5173",
                "http://localhost:8888",
                "http://localhost:8080",
                "https://dev-pick.com"
            )
            allowedMethods = listOf("GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS")
            allowedHeaders = listOf("*")
            allowCredentials = true
        }

        return UrlBasedCorsConfigurationSource().apply {
            registerCorsConfiguration("/**", configuration)
        }
    }
}
